File: Remove-Duplicates-from-Sorted-List-II/Remove-Duplicates-from-Sorted-List-II.py

‘’’
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        if head==None:
            return None
        if head.next==None:
            return head
        pprev=ListNode(-1)
        s=pprev
        p=head
        q=head.next
        pprev.next=head
        while q!=None:
            if p.val!=q.val:
                s=p
                p=q
                q=q.next
            else:
                while q!=None and q.val==p.val:
                    q=q.next
                if q==None:
                    s.next=None
                else:
                    p=q
                    q=q.next
                    s.next=p
        return pprev.next
 

File: Remove-Duplicates-from-Sorted-List-II/Remove-Duplicates-from-Sorted-List-II.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        if(!head || !head->next)
            return head;
        ListNode *p=NULL;
        ListNode *q=head;
        ListNode *phead=NULL;
        while(q)
        {
            if(!q->next || q->val!=q->next->val)
            {
                if(p)
                    p->next=q;
                else
                    phead=q;
                p=q;
            }
            else
            {
                while(q->next && q->val==q->next->val)
                    q=q->next;
            }
            q=q->next;
        }
        if(phead)
            p->next=q;
        return phead;
    }
};


 